x-pack/plugins/task_manager/server/lib/bulk_operation_buffer.ts

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { keyBy, map } from 'lodash';
+import { Subject } from 'rxjs';
+import { bufferWhen, filter } from 'rxjs/operators';
+import { either, Result, asOk, asErr, Ok, Err } from './result_type';
+
+export interface Entity {
+  id: string;
+}
+
+export interface OperationError<Input, ErrorOutput> {
+  entity: Input;
+  error: ErrorOutput;
+}
+
+export type OperationResult<Input, ErrorOutput, Output = Input> = Result<
+  Output,
+  OperationError<Input, ErrorOutput>
+>;
+
+export type Operation<Input, ErrorOutput, Output = Input> = (
+  entity: Input
+) => Promise<Result<Output, ErrorOutput>>;
+
+export type BulkOperation<Input, ErrorOutput, Output = Input> = (
+  entities: Input[]
+) => Promise<Array<OperationResult<Input, ErrorOutput, Output>>>;
+
+export function createBuffer<Input extends Entity, ErrorOutput, Output extends Entity = Input>(
+  bulkOperation: BulkOperation<Input, ErrorOutput, Output>
+): Operation<Input, ErrorOutput, Output> {
+  const flushBuffer = new Subject<void>();
+  const storeUpdateBuffer = new Subject<{
+    entity: Input;
+    onSuccess: (entity: Ok<Output>) => void;
+    onFailure: (error: Err<ErrorOutput>) => void;
+  }>();
+
+  storeUpdateBuffer
+    .pipe(
+      bufferWhen(() => flushBuffer),
+      filter((tasks) => tasks.length > 0)
+    )
+    .subscribe((entities) => {
+      const entityById = keyBy(entities, ({ entity: { id } }) => id);
+      bulkOperation(map(entities, 'entity'))
+        .then((results) => {
+          results.forEach((result) =>
+            either(
+              result,
+              (entity) => {
+                entityById[entity.id].onSuccess(asOk(entity));
+              },
+              ({ entity, error }: OperationError<Input, ErrorOutput>) => {
+                entityById[entity.id].onFailure(asErr(error));
+              }
+            )
+          );
+        })
+        .catch((ex) => {
+          entities.forEach(({ onFailure }) => onFailure(asErr(ex)));
+        });
+    });
+
+  return async function (entity: Input) {
+    return new Promise((resolve, reject) => {
+      // ensure we flush by the end of the "current" event loop tick
+      setImmediate(() => flushBuffer.next());
```

Comment:
Have we thought about using a time and size based approach (ex: when queue size is 100 or when 50ms has passed)? I'm wondering the effectiveness of relying on the even loop for tasks that don't complete immediately and rather within a time window.

Reply:
Yeah, this is doable.
I think 50ms is too long, as these are tasks that have already completed and are being staled by this batch, but we can make it configurable.
I'll look at configuring the batch size too... not sure how doable that will be in the Rx code... I'll investigate. 😬

Reply:
++ yeah the above is definitely example numbers.

Reply:
I'm not as worried about the time as the space - number of entries being updated.  The unbounded queue problem :-).  Seems to me this is very unlikely going to be a problem - isn't the max really `max_workers`?  Still would be nice to have some kind of limit, or check with a warning or such.

Reply:
I like that idea @pmuellr - I've now added an upper bound of the `max_workers` and also added a configurable flush duration which I'm defaulting to 50ms...

we'll see how it behaves

Reply:
It now allows us to set a longer buffer duration (50ms by default) and limits the buffer to the number of MAX_WORKERS.
It’s more complicated for sure, but I think it’s a good long term direction, it makes more sense.